feat: Introduce tedge http command (#3357)
thin-edge/thin-edge.io: docs/src/references/cli/tedge-http.md

---
title: "tedge http"
tags: [Reference, CLI]
sidebar_position: 6
---

# The tedge http command

A `tedge` sub command to interact with the HTTP services hosted on the device by the Cumulocity mapper and the agent:

- the [Cumulocity proxy](/references/cumulocity-proxy/)
- the [file transfer service](/references/file-transfer-service)

> - the entity store.

Author: @albinsuresh do we have some documentation on the HTTP API for device registration?

Reviewer: Yes. At `doc/src/operate/registration/register.md`

Author: I added the link removing the draft status of this page (a draft page cannot be linked to): https://github.com/thin-edge/thin-edge.io/pull/3357/commits/f71705349e51b269ab232631e3b7417334a342d8

This command uses `tedge config` to get the appropriate host, port and credentials to reach these local HTTP services.
So the same command can be used unchanged from the main device or a child device, with TLS or mTLS enabled or not.

```sh title="tedge http"
Send HTTP requests to local thin-edge HTTP servers

Usage: tedge http [OPTIONS] <COMMAND>

Commands:
  post    POST content to thin-edge local HTTP servers
  put     PUT content to thin-edge local HTTP servers
  get     GET content from thin-edge local HTTP servers
  delete  DELETE resource from thin-edge local HTTP servers
  help    Print this message or the help of the given subcommand(s)

Options:
      --config-dir <CONFIG_DIR>  [env: TEDGE_CONFIG_DIR, default: /etc/tedge]
      --debug                    Turn-on the DEBUG log level
      --log-level <LOG_LEVEL>    Configures the logging level
  -h, --help                     Print help (see more with '--help')
```

> ## Local HTTP services

Reviewer: ```suggestion
## Gateway HTTP services
```
Since these services are accessible remotely as well. Actually that word `thin-edge local` is used across this document to indicate that thin-edge gateway device providing those services. But the word `local` is a bit confusing, esp in the context of a child device that also has thin-edge installed on it. How about `thin-edge gateway HTTP services` or `thin-edge main HTTP services`?

`tedge http` forwards requests to the appropriate local HTTP service using the URL prefixes.

The requests are forwarded to the appropriate service depending on the URL prefix.

- URIs prefixed by `/c8y/` are forwarded to the [Cumulocity Proxy](/references/cumulocity-proxy/)

   ```sh title="Interacting with Cumulocity"
   tedge http get /c8y/inventory/managedObjects
   ```

- URIs starting with `/tedge/file-transfer/` are directed to the [File Transfer Service](/references/file-transfer-service)

>    ```sh title="Transfering files across child devices"

Reviewer: ```suggestion
   ```sh title="Transferring files to the main/gateway device"
```

Author: Done: https://github.com/thin-edge/thin-edge.io/pull/3357/commits/f71705349e51b269ab232631e3b7417334a342d8

   tedge http put /tedge/file-transfer/target.txt --file source.txt
   ```
  

> - URIs starting with `/tedge/entity-store` are related to the Entity Registration

Reviewer: ```suggestion
- URIs starting with `/tedge/entity-store` are directed to the [Entity Store Service](/operate/registration/register)
```

Author: done: https://github.com/thin-edge/thin-edge.io/pull/3357/commits/f71705349e51b269ab232631e3b7417334a342d8

>    ```sh title="Listing child-devices and services"

Reviewer: ```suggestion
   ```sh title="Listing all entities"
```

Author: done: https://github.com/thin-edge/thin-edge.io/pull/3357/commits/f71705349e51b269ab232631e3b7417334a342d8

   tedge http get /tedge/entity-store/v1/entities
   ```

## Configuration

> For `tedge http` to be used from the main device or any child device, with TLS or mTLS enabled or not,

Reviewer: ```suggestion
For `tedge http` to be used from the main device or any client device, with TLS or mTLS enabled or not,
```
Used the word `client` instead of child, as this tool can be used from any client device that has `tedge` installed.

Author: Done https://github.com/thin-edge/thin-edge.io/pull/3357/commits/f71705349e51b269ab232631e3b7417334a342d8

the host, port and credentials of the local %%te%% HTTP services
have to be properly configured on the main device as well as the child devices.

> ### On the box running the main agent

Reviewer: Optional:
```suggestion
### On the device running the main agent
```

Author: By box, I mean host or container.

Reviewer 2: We can leave it there for now, however the "device running" is used in more places in the docs (also to referring to the device in either main device or child device context), so we can create another PR to normalize the language. Below just shows a quick search comparing the two phrases:

"box running" - 3 times (only on **http-configuration.md** doc)
"device running" - 4 times over 3 different pages

Reviewer 2: Though I wouldn't be against changing "box/device running" to "host running"...

Author: Changed to *host*: https://github.com/thin-edge/thin-edge.io/pull/3357/commits/1055504f105d709da5bd7ecb315a01ec3bf6f3f4

on the HTTP services provided by the Cumulocity mapper ([Cumulocity proxy](/references/cumulocity-proxy/)).
This can be done along three security levels.

```sh title="Listening HTTP requests"
    c8y.proxy.bind.address  The IP address local Cumulocity HTTP proxy binds to. 
                            Example: 127.0.0.1
       c8y.proxy.bind.port  The port local Cumulocity HTTP proxy binds to. 
                            Example: 8001
```

```sh title="Enabling TLS aka HTTPS"
       c8y.proxy.cert_path  The file that will be used as the server certificate for the Cumulocity proxy. 
                            Example: /etc/tedge/device-certs/c8y_proxy_certificate.pem
        c8y.proxy.key_path  The file that will be used as the server private key for the Cumulocity proxy. 
                            Example: /etc/tedge/device-certs/c8y_proxy_key.pem
```

```sh title="Enforcing mTLS"
         c8y.proxy.ca_path  Path to a file containing the PEM encoded CA certificates that are trusted when checking incoming client certificates for the Cumulocity Proxy. 
                            Example: /etc/ssl/certs
```

> ### One all boxes

Reviewer: ```suggestion
### On all client boxes
```

Author: Just be aware that this has also to be configured on the main device.

Author: Done https://github.com/thin-edge/thin-edge.io/pull/3357/commits/f71705349e51b269ab232631e3b7417334a342d8

Reviewer 2: Partially addressed, but still renaming the "one" to "on", but this is superceded by this [comment](https://github.com/thin-edge/thin-edge.io/pull/3357#discussion_r1946425250)

The following `tedge config` settings control how child devices access the local HTTP services.
This has to be done in consistent way with the main agent and Cumulocity mapper settings.

```sh title="Reaching local HTTP services"
     c8y.proxy.client.host  The address of the host on which the local Cumulocity HTTP Proxy is running, used by the Cumulocity mapper. 
                            Examples: 127.0.0.1, 192.168.1.2, tedge-hostname
     c8y.proxy.client.port  The port number on the remote host on which the local Cumulocity HTTP Proxy is running, used by the Cumulocity mapper. 

>                             Example: 8001

Author: `http.client.host` and `http.client.port` are missing.

Author: added https://github.com/thin-edge/thin-edge.io/pull/3357/commits/f71705349e51b269ab232631e3b7417334a342d8

```

```sh title="Using TLS aka HTTPS"
              http.ca_path  Path to a directory containing the PEM encoded CA certificates that are trusted when checking incoming client certificates for the File Transfer Service. 
                            Example: /etc/ssl/certs
         c8y.proxy.ca_path  Path to a file containing the PEM encoded CA certificates that are trusted when checking incoming client certificates for the Cumulocity Proxy. 
                            Example: /etc/ssl/certs
```

```sh title="Using mTLS"
http.client.auth.cert_file  Path to the certificate which is used by the agent when connecting to external services. 

>  http.client.auth.key_file  Path to the private key which is used by the agent when connecting to external services. 

Author: There are no settings for `c8y.proxy.client.auth`. Am I missing something?

Reviewer 2: Yeah you're right, that is a bit unexpected. Maybe the tedge-agent uses the `http.client.auth.*` properties when interacting with with the file transfer server or the c8y proxy (as the property is called `http.client.auth*`), but still it would open up for misconfiguration if the user configured the c8y.proxy (server) to use a different certificate as the tedge-agent's file transfer service (e.g. when they are not sharing certificates).

- the [Cumulocity Proxy](/references/cumulocity-proxy/)
- the [File Transfer Service](/references/file-transfer-service/)
- the [Entity Store Service](/operate/registration/register/).

This command uses `tedge config` to get the appropriate host, port and credentials to reach these local HTTP services.
So the same command can be used unchanged from the main device or a child device, with TLS or mTLS enabled or not.

```sh title="tedge http"
Send HTTP requests to local thin-edge HTTP servers

Usage: tedge http [OPTIONS] <COMMAND>

Commands:
  post    POST content to thin-edge local HTTP servers
  put     PUT content to thin-edge local HTTP servers
  get     GET content from thin-edge local HTTP servers
  delete  DELETE resource from thin-edge local HTTP servers
  help    Print this message or the help of the given subcommand(s)

Options:
      --config-dir <CONFIG_DIR>  [env: TEDGE_CONFIG_DIR, default: /etc/tedge]
      --debug                    Turn-on the DEBUG log level
      --log-level <LOG_LEVEL>    Configures the logging level
  -h, --help                     Print help (see more with '--help')
```

> ## Thin-edge HTTP services

Reviewer 2: We should be using the variable to refer to the product name. And if you need to reference the header from another part of the docs, then you can give it an explicit anchor (`{#http-services}`).

```suggestion
## %%te%% HTTP services
```

Author: Done: https://github.com/thin-edge/thin-edge.io/pull/3357/commits/1055504f105d709da5bd7ecb315a01ec3bf6f3f4

on the HTTP services provided by the Cumulocity mapper ([Cumulocity proxy](/references/cumulocity-proxy/)).
This can be done along three security levels.

```sh title="Listening HTTP requests"
    c8y.proxy.bind.address  The IP address local Cumulocity HTTP proxy binds to. 
                            Example: 127.0.0.1
       c8y.proxy.bind.port  The port local Cumulocity HTTP proxy binds to. 
                            Example: 8001
```

```sh title="Enabling TLS aka HTTPS"
       c8y.proxy.cert_path  The file that will be used as the server certificate for the Cumulocity proxy. 
                            Example: /etc/tedge/device-certs/c8y_proxy_certificate.pem
        c8y.proxy.key_path  The file that will be used as the server private key for the Cumulocity proxy. 
                            Example: /etc/tedge/device-certs/c8y_proxy_key.pem
```

```sh title="Enforcing mTLS"
         c8y.proxy.ca_path  Path to a file containing the PEM encoded CA certificates that are trusted when checking incoming client certificates for the Cumulocity Proxy. 
                            Example: /etc/ssl/certs
```

> ### One all client hosts

Reviewer 2: I assume this should be `On` and not `One`...

```suggestion
### On all client hosts
```

The following `tedge config` settings control how client devices access the local HTTP services.
This has to be done in consistent way with the main agent and Cumulocity mapper settings.

```sh title="Reaching local HTTP services"

>           http.client.port  The port number on the remote host on which the File Transfer Service HTTP server is running. 
>                             Example: 8000
>           http.client.host  The address of the host on which the File Transfer Service HTTP server is running. 
>                             Examples: 127.0.0.1, 192.168.1.2, tedge-hostname
>      c8y.proxy.client.host  The address of the host on which the local Cumulocity HTTP Proxy is running, used by the Cumulocity mapper. 

Reviewer 2: Some minor indentation fixes

```suggestion
     http.client.port  The port number on the remote host on which the File Transfer Service HTTP server is running. 
                            Example: 8000
     http.client.host  The address of the host on which the File Transfer Service HTTP server is running. 
                            Examples: 127.0.0.1, 192.168.1.2, tedge-hostname
     c8y.proxy.client.host  The address of the host on which the local Cumulocity HTTP Proxy is running, used by the Cumulocity mapper. 
```

Author: I kept the alignment used by `tedge config list --doc` i.e. centered.

Reviewer 2: ah yeah you're right, sorry I thought it was left-aligned.
